### PR TITLE
[feat] trainer: make dataloader throughput configurable

### LINKED ISSUE #332: [feature] Make DataLoader throughput tunable from config
## Motivation

During VLA/VLM training, DataLoader throughput options are currently hard-coded in the codebase, for example `num_workers=4` in the VLA and VLM dataloader builders. This is inconvenient when training across different machines, storage backends, and datasets, because users need to edit source files directly just to tune input-pipeline performance.

In addition, the current training loop exposes raw data/model timing but does not provide higher-level throughput indicators, so it is harder to tell whether a run is input-bound or model-bound when comparing configurations.

## Better idea

Expose common DataLoader throughput options through dataset config fields:

- `num_workers`
- `pin_memory`
- `persistent_workers`
- `prefetch_factor`
- `drop_last`
- `timeout`

and add lightweight benchmark logging during training:

- `timing/step`
- `throughput/samples_per_sec`
- `throughput/model_samples_per_sec`
- `throughput/data_wait_ratio`
- GPU memory allocated / reserved

This would keep existing defaults backward-compatible while allowing reproducible throughput tuning through YAML/CLI config instead of code edits.

## Expected benefit

- Easier performance tuning on different hardware and storage environments
- More reproducible training configs
- Faster diagnosis of dataloader bottlenecks
- Shared visibility into throughput across VLA-only, VLM-only, and co-training workflows

### PR DESCRIPTION
**Pull request by Travor278**
## Motivation

Closes #332

Thanks for the clear guidance in the issue. This PR keeps the scope focused on making DataLoader throughput tunable from config, while keeping the current training behavior unchanged for existing YAML files.

Before this change, VLA/VLM DataLoader options such as worker count were hard-coded in the builders. That made throughput tuning awkward across different machines, storage backends, and datasets, because users had to edit source code for what should be a training configuration choice. The training loop also logged raw data/model timing, but did not group the higher-level throughput and GPU memory signals that make input-pipeline bottlenecks easier to compare.

## Changes

- Added dataset-level DataLoader options for both `datasets.vla_data` and `datasets.vlm_data`:
  - `num_workers`
  - `pin_memory`
  - `persistent_workers`
  - `prefetch_factor`
  - `drop_last`
  - `timeout`
- Preserved the previous hard-coded defaults for backward compatibility:
  - `num_workers: 4`
  - `pin_memory: false`
  - `persistent_workers: false`
  - `prefetch_factor: 2`
  - `drop_last: false`
  - `timeout: 0`
- Added a shared DataLoader options helper used by both VLA and VLM builders.
- Handles `num_workers: 0` safely by omitting `prefetch_factor`.
- Rejects `persistent_workers: true` with `num_workers: 0`, which is an invalid PyTorch combination.
- Added grouped training metrics across VLA-only, VLM-only, and co-training:
  - `timing/step`
  - `throughput/samples_per_sec`
  - `throughput/model_samples_per_sec`
  - `throughput/data_wait_ratio`
  - `memory/gpu_allocated_gb`
  - `memory/gpu_reserved_gb`
- Kept existing `timing/data` and `timing/model` metrics.
- Added VLM-only timing so all three training modes report consistently.
- Added core config defaults and documentation for YAML/CLI overrides.

## Testing

- [x] Ran `black --check` on the touched Python files - passes
- [x] Ran `ruff check` on the touched Python files - passes
- [x] Ran focused unit tests - `Ran 12 tests ... OK`
- [x] Ran `git diff --check` - passes
- [x] Verified touched module imports on Windows Python 3.11
- [x] Verified CUDA memory metric shape on RTX 5070 Laptop GPU
- [x] Verified real DataLoader iteration with `num_workers=1`
- [x] Verified WSL Ubuntu 22.04 training environment:
  - Python 3.10.12
  - CUDA Toolkit 12.8 / `nvcc` 12.8
  - PyTorch 2.7.1+cu128
  - DeepSpeed 0.16.9
  - RTX 5070 Laptop GPU
- [x] Ran real VLA training smoke for 3 steps with Qwen3.5-0.8B + RoboChallenge Table30v2 `shred_paper`
- [x] Ran real VLA training smoke for 3 steps with configured worker throughput options:
  - `num_workers=2`
  - `pin_memory=true`
  - `persistent_workers=true`
  - `prefetch_factor=4`

Commands run:

```bash
D:\Dev\conda-envs\py312\python.exe -m black --check <touched-python-files>
D:\Dev\conda-envs\py312\python.exe -m ruff check <touched-python-files>
D:\Dev\conda-envs\py311\python.exe -m unittest tests.test_dataloader_throughput
git diff --check
```

Real training smoke command shape:

```bash
accelerate launch \
  --config_file starVLA/config/deepseeds/deepspeed_zero2.yaml \
  --num_processes 1 \
  starVLA/training/train_starvla.py \
  --config_yaml ./examples/RoboChallenge_table30v2/train_files/starvla_qwenoft_robochallenge_table30v2.yaml \
  --framework.qwenvl.base_vlm <local-qwen3.5-0.8b> \
  --datasets.vla_data.data_root_dir <local-robochallenge-table30v2> \
  --datasets.vla_data.per_device_batch_size 1 \
  --datasets.vla_data.num_workers 2 \
  --datasets.vla_data.pin_memory true \
  --datasets.vla_data.persistent_workers true \
  --datasets.vla_data.prefetch_factor 4 \
  --trainer.max_train_steps 3 \
  --trainer.freeze_modules qwen_vl_interface.model
```

Observed training log excerpt:

```text
Step 3:
timing/step: 0.2509
throughput/samples_per_sec: 3.9861
throughput/model_samples_per_sec: 3.9886
throughput/data_wait_ratio: 0.0006
memory/gpu_allocated_gb: 1.7419
memory/gpu_reserved_gb: 3.7578
```

## Breaking Changes

None. Existing YAML files remain compatible because missing DataLoader fields fall back to the previous hard-coded behavior.

## Checklist

- [x] Branch is based on latest `starVLA_dev`
- [x] PR targets `starVLA_dev`
- [x] Scope stays focused on DataLoader throughput configuration and training metrics
- [x] Defaults preserve previous behavior
- [x] CLI overrides are documented
- [x] Unit tests cover defaults, overrides, invalid combinations, sample counting, throughput math, and CUDA/no-CUDA memory metric shape
- [x] Real training smoke runs for N steps without errors
- [x] No unrelated refactors or drive-by formatting
- [x] No secrets, API keys, datasets, checkpoints, or generated smoke outputs are committed

### CODE REVIEW

**Comment by JinhuiYE** [comment]
Thank you for your PR. Sorry for the delay — the main team has final exams this week, so reviewers will start being processed after Wednesday (5/20).

**Comment by Travor278** [comment]
Thanks for the update, and no worries at all. Best of luck to everyone on the team with finals this week. I’ll keep the PR stable for review, and I can add one more follow-up note once I have access to official RoboChallenge remote scoring.

In the meantime, I ran an extra local end-to-end validation pass for this branch today before review starts.

Validation checklist:

- [x] `black --check` / `ruff check` on the touched Python files
- [x] focused `unittest` coverage for DataLoader kwargs, `num_workers=0`, invalid `persistent_workers`, sample counting, throughput math, and CUDA/no-CUDA memory metric shape
- [x] 100-step local VLA training run on WSL Ubuntu 22.04 with CUDA-visible PyTorch + DeepSpeed, using real RoboChallenge Table30v2 `shred_paper` data
- [x] exercised the new DataLoader override path during training: `num_workers=2`, `pin_memory=true`, `persistent_workers=true`, `prefetch_factor=4`
- [x] verified the new grouped metrics appear during training: `timing/step`, `throughput/samples_per_sec`, `throughput/model_samples_per_sec`, `throughput/data_wait_ratio`, `memory/gpu_allocated_gb`, `memory/gpu_reserved_gb`
- [x] loaded the 100-step checkpoint with the RoboChallenge policy bridge and ran local inference self-test: returned `(8, 8)` actions, avg latency about `124.6 ms` over 3 runs
- [x] ran a local RoboChallenge mock-server loop with the produced checkpoint: `get_state -> policy inference -> post_actions` completed for 5 iterations, action shape stayed `[8, 8]`

Local run details:

- Training checkpoint: `steps_100_pytorch_model.pt`
- 100-step training completed successfully and saved both checkpoint + final model
- Mock-server benchmark summary: avg state fetch `6.64 ms`, avg inference `331.61 ms` including first-call warmup, avg post `2.91 ms`; after warmup inference was around `114-128 ms`

One note: I have not run the official remote RoboChallenge scoring yet because that requires a RoboChallenge account token and submission id. I plan to add a follow-up validation note once those are available. The local mock-server path does exercise the same request/response shape and action posting path, but it is not a substitute for the official remote score.

**Comment by Travor278** [comment]
Small follow-up update while review is pending:

- Synced this branch with the latest `origin/starVLA_dev` and resolved the merge conflicts that appeared after the recent upstream changes.
- Kept the new upstream VLM/VLN dataloader changes, and re-applied this PR's configurable DataLoader kwargs on top of them.
- The PR is now reported as `MERGEABLE` by GitHub again.
- Re-ran focused checks after resolving the conflicts:
  - `python -m unittest tests.test_dataloader_throughput` - 12 tests passed
  - `ruff check` on the PR core files passed
  - `ruff check --select F821 starVLA/dataloader/vlm_datasets.py` passed
  - `py_compile` on the touched dataloader files passed
  - `git diff --check` passed

I also uploaded the 100-step RoboChallenge checkpoint used for local validation here:

https://huggingface.co/Travor278/starvla-qwen35-08b-shred-scrap-paper-100step

Official RoboChallenge remote scoring is still waiting on the platform side: the submitted `shred_scrap_paper` run is still `pending` with no robot assigned yet. I will add another note once the platform moves it to a runnable state.

**Comment by yeahjack** [comment]
@codex review
